crates/circuit/src/dag_circuit.rs

```diff
@@ -7095,6 +6965,273 @@ impl DAGCircuit {
+    /// be added continuously into the [DAGCircuit].
+    pub fn new(dag: DAGCircuit) -> DAGCircuitConcat {
+        let num_qubits = dag.num_qubits();
+        let num_clbits = dag.num_clbits();
+        let num_vars = dag.num_vars();
+        Self {
+            dag,
+            last_qubits: (0..num_qubits).map(|_| None).collect(),
+            last_clbits: (0..num_clbits).map(|_| None).collect(),
+            last_vars: (0..num_vars).map(|_| None).collect(),
+        }
+    }
+
+    /// Finishes up the changes by re-connecting all of the ouput nodes back to the last
+    /// recorded nodes.
+    pub fn end(mut self) -> DAGCircuit {
+        // Re-connects all of the output nodes with their respective last nodes.
+        // Add the output_nodes back to qargs
+        for (qubit, node) in self
+            .last_qubits
+            .into_iter()
+            .enumerate()
+            .filter_map(|(qubit, node)| node.map(|node| (qubit, node)))
+        {
+            let output_node = self.dag.qubit_io_map[qubit][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Qubit(Qubit(qubit as u32)));
+        }
+
+        // Add the output_nodes back to cargs
+        for (clbit, node) in self
+            .last_clbits
+            .into_iter()
+            .enumerate()
+            .filter_map(|(clbit, node)| node.map(|node| (clbit, node)))
+        {
+            let output_node = self.dag.clbit_io_map[clbit][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Clbit(Clbit(clbit as u32)));
+        }
+
+        // Add the output_nodes back to vars
+        for (var, node) in self
+            .last_vars
+            .into_iter()
+            .enumerate()
+            .filter_map(|(var, node)| node.map(|node| (var, node)))
+        {
+            let output_node = self.dag.var_io_map[var][1];
+            self.dag
+                .dag
+                .add_edge(node, output_node, Wire::Var(Var(var as u32)));
+        }
+        self.dag
+    }
+
+    /// Applies a new operation to the back of the circuit. This variant works with non-owned bit indices.
+    pub fn apply_operation_back(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qubits: Option<Cow<[Qubit]>>,
+        clbits: Option<Cow<[Clbit]>>,
+        params: Option<Box<SmallVec<[Param; 3]>>>,
+        label: Option<String>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+    ) -> NodeIndex {
+        let instruction = self.pack_instruction(
+            op,
+            qubits,
+            clbits,
+            params,
+            label,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+        );
+        self.push_instruction_back(py, instruction)
+    }
+
+    /// Pushes a valid [PackedInstruction] to the back ot the circuit.
+    pub fn push_instruction_back(&mut self, py: Python<'_>, inst: PackedInstruction) -> NodeIndex {
```

Comment:
It looks like this function duplicates a lot of what's happening in `DAGCircuit::push_back`, with the exception of course of adding back the output edges.

Would it be possible to create a private function on `DAGCircuit` that we could call both from the implementation of `DAGCircuit::push_back` and `DAGCircuitConcat::push_instruction_back` that does the shared bits of logic?

Reply:
I will add a `_push_back` method that does the first steps, since in the end we don't perform the same operations for modifying the nodes.